packages/actor-rdf-resolve-quad-pattern-federated/lib/FederatedQuadSource.ts

```diff
@@ -170,30 +203,111 @@ export class FederatedQuadSource implements IQuadSource {
     return sourceId;
   }
 
+
+
+  checkPushEmptyPattern = (md: Record<string,any> | undefined, source: IDataSource,
+                                 pattern: RDF.BaseQuad | undefined, lastMetadata?: Record<string, any>) => {
+    if (this.skipEmptyPatterns && !md && pattern && !this.isSourceEmpty(source, pattern)) {
+      this.emptyPatterns.get(source)!.push(pattern);
+    }
+  }
+
   public match(subject: RDF.Term, predicate: RDF.Term, object: RDF.Term, graph: RDF.Term): AsyncIterator<RDF.Quad> {
-    // Counters for our metadata
-    const metadata: Record<string, any> = { totalItems: 0 };
-    let remainingSources = this.sources.length;
-
-    // Anonymous function to handle totalItems from metadata
-    const checkEmitMetadata = (currentTotalItems: number, source: IDataSource,
-      pattern: RDF.BaseQuad | undefined, lastMetadata?: Record<string, any>): void => {
-      if (this.skipEmptyPatterns && !currentTotalItems && pattern && !this.isSourceEmpty(source, pattern)) {
-        this.emptyPatterns.get(source)!.push(pattern);
-      }
-      if (!remainingSources) {
-        if (lastMetadata && this.sources.length === 1) {
-          // If we only had one source, emit the metadata as-is.
-          it.setProperty('metadata', lastMetadata);
-        } else {
-          it.setProperty('metadata', metadata);
-        }
+
+    const metadataCallbacks =
```

Comment:
Not sure I understand what is going on here, but I suspect you may be overcomplicating things.

If I understand your code correctly, your iterating over the sources a second time, and invoking `mediatorResolveQuadPattern` manually again to obtain the metadata.
This is not something we want to do, because we're otherwise calling expensive querying code twice, thereby making queries twice as slow, which is not what we want.

I would suggest analyzing the original code again wrt metadata handling, and modifying the parts where `totalItems` are hardcoded to invoke the new metadata-aggregate bus instead.